### DISCUSSION TITLE
Reacting to file changes?

### DISCUSSION THREAD

**Discussion by mpizenberg** [woelper/oculante#241] [Ideas]
For a some example code on how to use some industrial camera API, I'm repeatedly saving the captured image into the same image file. Granted, this isn't the best way to show images ^^. But it felt convenient with an image viewer that was able to watch file changes and reload the image. I'm just wondering if watching file changes to reload the current image was something possible with oculante?

I'll take the opportunity to say great job with oculante!

**Comment by woelper** [comment]
Hey! Thanks for the suggestion - a really great idea. Sorry for answering so late, I have to get used to the fact that people actually write items here, and I don't see all the notifications in time :) 

Thanks for the kind words, I'll add a ticket for it!

**Comment by woelper** [comment]
A couple of follow-up questions:

- Should the image be reset on change? For example if you zoomed in, should it revert to the original size?
- Should there be a notification that the image has been reloaded?
- Should this be optional, perhaps an option in settings or default? (I guess some image viewers reload as default)

**Reply by mpizenberg** [reply]
- reset on change: I'd guess keeping the current user setting would be useful. I'm often setting it to keeping current zoom because it makes comparing specific areas of related images easier.
- notification: I'd vote for no (at least for proper OS notifications) but if it's just withing the app, it might be useful indeed to have a small indicator that the underlying image changed. Not crucial but can help debugging to differentiate "the image did not change" and "the content of the image did not change".
- optional reload: I don't know. I think reloading makes sense as a default, but that's my use case so I wouldn't assume it's the same for others.

**Reply by woelper** [reply]
There is an initial implementation now done in a branch. I just need to find a good balance how to trigger infrequent updates. Oculante does not do anything currently when it is idle and unfocused, so you'd have to click on the window or at least hover over it to see changes. I'll see if I can change that.

**Reply by mpizenberg** [reply]
That's a very good start!

In my use case, the changes happen as I enter line return in the cli program so the focus is in the terminal and not on the image viewer.

**Reply by woelper** [reply]
I have looked at it more today, and there does not seem to be an easy way to wake up the app if there is no input. Unless I have a different idea, I will probably go with reloading if the image has focus and provide an additional way to force the app to redraw continuously. That way you could set it up the way you need.

**Reply by mpizenberg** [reply]
If that can help, I've built a file watcher for some other project of mine. When one of the files watched change, it recompiles them. I imagine something similar could be done? watch if the current image changes, re-decode it.

Here is the watching part in my project: https://github.com/mpizenberg/elm-test-rs/blob/d9e0c921c986311f80fdfccc7d2cd9502b9a9520/src/project.rs#L54

**Reply by woelper** [reply]
That is very kind of you, thanks!

The issue is not getting the file change, but working withing the limitations I have - as far as I know, there is no way to wake up the engine's "game loop" externally, so new changes will only be picked up once the application receives input. I've added the option to redraw every frame in settings, which should provide you with the functionality you need until I come up with a better, way. If you like, give the new release a spin and check if the points above have been addressed!

**Reply by mpizenberg** [reply]
Ah right, I've forgotten that GUIs are hard.

Thanks! I'll try it out during the week.
